_posts/2015/2015-12-08-wjax2015-microservices.md

```diff
@@ -0,0 +1,76 @@
+
+However a few of these new software design concepts are meaningful and solve current software development problems.
+One of them is the idea of microservices.
+The participants of the W-JAX 2015 conference talked about this idea in many sessions.
+
+## When do we need microservices
+
+Developing software until now often had the idea of developing a [monolith](https://en.wikipedia.org/wiki/Monolithic_system).
+With a monolith we build a software in one application with one database and (mostly) one programming language.
+To test the complete software will entail huge tests.
+
+With a few millions of lines of code developing and testing this big monolith will be a huge problem.
+If the developer wants to invest more time in service instead of in extending his code they should think about switching to a microservice system.
+This will also help to avoid using always the same patterns and requires a code generator to automatically create static structure.
+
+## The idea of microservices
+
+Basically, the idea of using microservices is to encapsulate every part of the software into a separate service.
+Every service can run on its own and use its own de-central database.
+If one service crashes the whole system can work without failing.
+It is possible to run more instances of this service to scale it in case of being required more often.
+
+To structure a monolith into small microservices, it is important to find a way to slice the software code into small parts.
+The following checklist shows some important remarks:
+
+- To be used in a way to avoid big communication.
+If two components have to communicate very often both should be in the same microservice.
+- Should not influence another microservice.
+- Every microservice can stand alone without knowing other components of the software.
+- Complete microservice should be handled by only one team.
+If the sliced software component cannot be serviced by one team alone, slice it again.
+
+## Challenges and hints
+
+Microservices will put the weight of complexity from algorithm complexity to communication complexity.
+The dependencies of each software part is now moved to the way of communication.
+
+A second big challenge can be the [Two-Phase Commit](http://www.enterpriseintegrationpatterns.com/ramblings/18_starbucks.html) and data consistency.
+Let's imagine we want to transfer money from one to another bank account.
+The monolith would have a controller layer which sends a request to the model layout.
+This layer will transfer the money from one to another account instantly.
+In a microservice world every bank account can be in another microservice.
+Microservice 1 will send the money to microservice 2 and decrease the money in its own database.
+In this moment the money is not existent: neither in microservice 1 nor in microservice 2.
```

Comment:
The talk "Jave EE meets Microservices" covered these topics and explained, what eventual consistency is and that you cannot guarantee immediate consistency in distributed environments. See also http://www.enterpriseintegrationpatterns.com/ramblings/18_starbucks.htm